src/ILLink.Shared/TypeSystemProxy/MethodProxy.cs

```diff
@@ -13,17 +13,51 @@ namespace ILLink.Shared.TypeSystemProxy
 		// Currently this only needs to work on non-nested, non-generic types.
 		// The format of the fullTypeName parameter is 'namespace.typename', so for example 'System.Reflection.Assembly'
 		internal partial bool IsDeclaredOnType (string fullTypeName);
-		internal partial bool HasParameters ();
+
+		/// <summary>
+		/// Returns the number of the parameters in the 'parameters' metadata section. This should map directly to the number of parameters in the C# source declaration as well.
+		/// </summary>
+		internal partial int GetMetadataParametersCount ();
+
+		/// <summary>
+		/// Returns true if the method has parameters in the 'paramters' metadata section (i.e. has parameters besides the implicit 'this' parameter)
+		/// </summary>
+		internal partial bool HasMetadataParameters ();
+
+		/// <summary>
+		/// Returns the number of parameters that are passed to the method in IL (including the implicit 'this' parameter).
+		/// In pseudocode: <code>method.HasImplicitThis() ? 1 + MetadataParametersCount : MetadataParametersCount;</code>
+		/// </summary>
 		internal partial int GetParametersCount ();
-		internal bool HasParametersCount (int parameterCount) => GetParametersCount () == parameterCount;
+
+		/// <summary>
+		/// Returns a List of <see cref="ParameterProxy"/> representing the parameters the method takes, including the implicit 'this' parameters.
+		/// </summary>
+		internal partial ParameterCollection GetParameters ();
+
+		/// <summary>
+		/// Returns the ParameterProxy corresponding to the parameter at <paramref name="index"/>.
+		/// <paramref name="index"/> is the index of the parameters as they are passed to the method, with 0 being the implicit this parameter if it exists.
+		/// See <see cref="ParameterIndex"/> for more info.
+		/// </summary>
+		internal partial ParameterProxy? GetParameter (ParameterIndex index);
```

Comment:
Do we actually use the fact that it can return `null` if the parameter index is out of range?
I would prefer we throw in that case.

Reply:
We use it in MethodProxy.HasParameterOfType() where we return false if the index is out of range and GetParameter, but we could have an out of range check in HasParameterOfType() and throw in GetParameter.

Reply:
I changed it to throw and added a bounds check in HasParameterOfType()